language/diem-framework/modules/NetworkIdentity.move

```diff
@@ -44,93 +46,91 @@ module DiemFramework::NetworkIdentity {
+            Errors::invalid_state(ENETWORK_ID_DOESNT_EXIST)
+        );
         *&borrow_global<NetworkIdentity>(account_addr).identities
     }
-
     spec get {
         aborts_if !exists<NetworkIdentity>(account_addr);
+        ensures result == global<NetworkIdentity>(account_addr).identities;
     }
 
     /// Update and create if not exist `NetworkIdentity`
     public fun add_identities(account: &signer, to_add: vector<vector<u8>>) acquires NetworkIdentity {
         let num_to_add = Vector::length(&to_add);
-        assert(num_to_add > 0, ENETWORK_ID_NO_INPUT);
+        assert(num_to_add > 0, Errors::invalid_argument(ENETWORK_ID_NO_INPUT));
 
         if (!exists<NetworkIdentity>(Signer::address_of(account))) {
             initialize_network_identity(account);
         };
         let identity = borrow_global_mut<NetworkIdentity>(Signer::address_of(account));
         let identities = &mut identity.identities;
 
-        assert(Vector::length(identities) + num_to_add <= MAX_ADDR_IDENTITIES, Errors::limit_exceeded(ENETWORK_ID_LIMIT_EXCEEDED));
-
-        let i = 0;
-        let has_change = false;
-        while (i < num_to_add) {
-           has_change = has_change || add_identity(identities, *Vector::borrow(&to_add, i));
-           i = i + 1;
-        };
-
+        assert(
+            Vector::length(identities) + num_to_add <= MAX_ADDR_IDENTITIES,
+            Errors::limit_exceeded(ENETWORK_ID_LIMIT_EXCEEDED)
+        );
 
+        let has_change = add_members_internal(identities, &to_add);
         if (has_change) {
             Event::emit_event(&mut identity.identity_change_events, NetworkIdentityChangeNotification {
                 identities: *&identity.identities,
                 time_rotated_seconds: DiemTimestamp::now_seconds(),
             });
         }
     }
-
     spec add_identities {
         let account_addr = Signer::spec_address_of(account);
-        let num_identities = len(global<NetworkIdentity>(account_addr).identities);
-        // aborts_if len(to_add) == 0;
-        // aborts_if num_identities + len(to_add) > MAX_ADDR_IDENTITIES;
-        // aborts_if !exists<NetworkIdentity>(account_addr);
-        modifies global<NetworkIdentity>(account_addr);
-        invariant exists<NetworkIdentity>(account_addr);
-        invariant num_identities <= MAX_ADDR_IDENTITIES;
-    }
-
-    /// Adds an identity and returns true if a change was made
-    fun add_identity(identities: &mut vector<vector<u8>>, to_add: vector<u8>): bool {
-        if (!Vector::contains(identities, &to_add)) {
-            Vector::push_back(identities, to_add);
-            true
+        let prior_identities = if (exists<NetworkIdentity>(account_addr)) {
+            global<NetworkIdentity>(account_addr).identities
         } else {
-            false
-        }
-    }
+            vec()
+        };
+        let has_change = (exists i in 0..len(to_add): !contains(prior_identities, to_add[i]));
```

Comment:
Instead of 0..len(vec) you can use range(vec). Or even better quantify over the elements in the vector:

`let has_change = (exists x in to_add: !contains(prior_identities, x)`